xbmc/utils/BitstreamReader.cpp

```diff
@@ -0,0 +1,128 @@
+/*
+*      Copyright (C) 2017 Team XBMC
+*      http://xbmc.org
+*
+*  This Program is free software; you can redistribute it and/or modify
+*  it under the terms of the GNU General Public License as published by
+*  the Free Software Foundation; either version 2, or (at your option)
+*  any later version.
+*
+*  This Program is distributed in the hope that it will be useful,
+*  but WITHOUT ANY WARRANTY; without even the implied warranty of
+*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+*  GNU General Public License for more details.
+*
+*  You should have received a copy of the GNU General Public License
+*  along with XBMC; see the file COPYING.  If not, see
+*  <http://www.gnu.org/licenses/>.
+*
+*/
+
+#include "BitstreamReader.h"
+
+CBitstreamReader::CBitstreamReader(const uint8_t *buf, int len)
+  : buffer(buf)
+  , start(buf)
+  , offbits(0)
+  , length(len)
+  , oflow(0)
+{
+}
+
+uint32_t CBitstreamReader::ReadBits(int nbits)
+{
+  int i, nbytes;
+  uint32_t ret = 0;
+  const uint8_t *buf;
+
+  buf = buffer;
+  nbytes = (offbits + nbits) / 8;
+
+  if (((offbits + nbits) % 8) > 0)
+    nbytes++;
+
+  if ((buf + nbytes) > (start + length))
+  {
+    oflow = 1;
+    return 0;
+  }
+  for (i = 0; i<nbytes; i++)
+    ret += buf[i] << ((nbytes - i - 1) * 8);
+
+  i = (4 - nbytes) * 8 + offbits;
+
+  ret = ((ret << i) >> i) >> ((nbytes * 8) - nbits - offbits);
+
+  offbits += nbits;
+  buffer += offbits / 8;
+  offbits %= 8;
+
+  return ret;
+}
+
+void CBitstreamReader::SkipBits(int nbits)
+{
+  offbits += nbits;
+  buffer += offbits / 8;
+  offbits %= 8;
+
+  if (buffer > (start + length))
+    oflow = 1;
+}
+
+uint32_t CBitstreamReader::GetBits(int nbits)
+{
+  int i, nbytes;
+  uint32_t ret = 0;
+  const uint8_t *buf;
+
+  buf = buffer;
+  nbytes = (offbits + nbits) / 8;
+
+  if (((offbits + nbits) % 8) > 0)
+    nbytes++;
+
+  if ((buf + nbytes) > (start + length))
+  {
+    oflow = 1;
+    return 0;
+  }
+  for (i = 0; i<nbytes; i++)
+    ret += buf[i] << ((nbytes - i - 1) * 8);
+
+  i = (4 - nbytes) * 8 + offbits;
+
+  ret = ((ret << i) >> i) >> ((nbytes * 8) - nbits - offbits);
```

Comment:
Up to here, this function duplicates ReadBits. Consider moving the common parts to a separate inline function.